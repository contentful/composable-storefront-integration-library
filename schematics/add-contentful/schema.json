{
    "$schema": "http://json-schema.org/schema",
    "$id": "ContentfulSchematics",
    "title": "Contentful Schematics",
    "type": "object",
    "properties": {
      "project": {
        "type": "string",
        "description": "The name of the project.",
        "$default": {
          "$source": "projectName"
        }
      },
      "debug": {
        "description": "Display additional details during the running process.",
        "type": "boolean",
        "default": false
      },
      "lazy": {
        "type": "boolean",
        "description": "Lazy load the smartedit features.",
        "default": true
      },
      "features": {
        "type": "array",
        "uniqueItems": true,
        "default": ["Contentful"]
      },
      "storefrontPreviewRoute": {
        "type": "string",
        "description": "The storefront preview route. (e.g. cx-preview)",
        "x-prompt": "[Contentful] What is the storefront preview route? (optional)"
      },
      "allowOrigin": {
        "type": "string",
        "description": "The URL for which to allow origin. (e.g. localhost:9002)",
        "x-prompt": "[Contentful] URL for which to allow origin? (optional)"
      }
    },
    "required": []
  }
  