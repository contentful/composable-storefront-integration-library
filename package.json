{
    "name": "spartacus-contentful-feature",
    "version": "1.0.0",
    "description": "Contentful CMS feature library for Spartacus",
    "license": "Apache-2.0",
    "homepage": "https://github.com/contentful/spartacus-contentful-feature",
    "repository": "https://github.com/contentful/spartacus-contentful-feature.git",
    "keywords": [
      "spartacus",
      "framework",
      "storefront",
      "contenetful"
    ],
    "scripts": {
      "build:schematics": "npm run clean:schematics && ../../node_modules/.bin/tsc -p ./tsconfig.schematics.json",
      "clean:schematics": "../../node_modules/.bin/rimraf --glob \"schematics/**/*.js\" \"schematics/**/*.js.map\" \"schematics/**/*.d.ts\"",
      "test:schematics": "npm --prefix ../../projects/schematics/ run clean && npm run clean:schematics && ../../node_modules/.bin/jest --config ./jest.schematics.config.js"
    },
    "dependencies": {
      "tslib": "^2.6.2"
    },
    "peerDependencies": {
      "@angular-devkit/schematics": "^17.0.5",
      "@angular/common": "^17.0.5",
      "@angular/core": "^17.0.5",
      "@spartacus/core": "2211.27.0",
      "@spartacus/schematics": "2211.27.0",
      "rxjs": "^7.8.0"
    },
    "publishConfig": {
      "access": "public"
    },
    "engines": {
      "node": "^18.20.0 || ^20.12.0"
    },
    "schematics": "./schematics/collection.json"
  }
  