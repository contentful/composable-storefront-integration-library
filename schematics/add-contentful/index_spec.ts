/// <reference types="jest" />

import {
    SchematicTestRunner,
    UnitTestTree,
  } from '@angular-devkit/schematics/testing';
  import {
    Schema as ApplicationOptions,
    Style,
  } from '@schematics/angular/application/schema';
  import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
  import {
    CONTENTFUL_FEATURE_NAME,
    SPARTACUS_SCHEMATICS,
    SPARTACUS_CONTENTFUL,
    SpartacusOptions,
    SpartacusContentfulOptions,
    contentfulFeatureModulePath,
  } from '@spartacus/schematics';
  import * as path from 'path';
  import { peerDependencies } from '../../package.json';
  
  const collectionPath = path.join(__dirname, '../collection.json');
  
  describe('Spartacus SmartEdit schematics: ng-add', () => {
    const schematicRunner = new SchematicTestRunner(
      SPARTACUS_CONTENTFUL,
      collectionPath
    );
  
    let appTree: UnitTestTree;
  
    const workspaceOptions: WorkspaceOptions = {
      name: 'workspace',
      version: '0.5.0',
    };
  
    const appOptions: ApplicationOptions = {
      name: 'schematics-test',
      inlineStyle: false,
      inlineTemplate: false,
      style: Style.Scss,
      skipTests: false,
      projectRoot: '',
      standalone: false,
    };
  
    const spartacusDefaultOptions: SpartacusOptions = {
      project: 'schematics-test',
      lazy: true,
      features: [],
    };
  
    const libraryNoFeaturesOptions: SpartacusContentfulOptions = {
      project: 'schematics-test',
      lazy: true,
      features: [],
    };
  
    const contentfulFeatureOptions: SpartacusContentfulOptions = {
      ...libraryNoFeaturesOptions,
      features: [CONTENTFUL_FEATURE_NAME],
    };
  
    beforeEach(async () => {
      schematicRunner.registerCollection(
        SPARTACUS_SCHEMATICS,
        '../../projects/schematics/src/collection.json'
      );
  
      appTree = await schematicRunner.runExternalSchematic(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      );
  
      appTree = await schematicRunner.runExternalSchematic(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      );
  
      appTree = await schematicRunner.runExternalSchematic(
        SPARTACUS_SCHEMATICS,
        'ng-add',
        { ...spartacusDefaultOptions, name: 'schematics-test' },
        appTree
      );
    });
  
    describe('Without features', () => {
      beforeEach(async () => {
        appTree = await schematicRunner.runSchematic(
          'ng-add',
          libraryNoFeaturesOptions,
          appTree
        );
      });
  
      it('should not create any of the feature modules', () => {
        expect(appTree.exists(contentfulFeatureModulePath)).toBeFalsy();
      });
  
      it('should install necessary Spartacus libraries', () => {
        const packageJson = JSON.parse(appTree.readContent('package.json'));
        let dependencies: Record<string, string> = {};
        dependencies = { ...packageJson.dependencies };
        dependencies = { ...dependencies, ...packageJson.devDependencies };
  
        for (const toAdd in peerDependencies) {
          // skip the SPARTACUS_SCHEMATICS, as those are added only when running by the Angular CLI, and not in the testing environment
          if (
            !peerDependencies.hasOwnProperty(toAdd) ||
            toAdd === SPARTACUS_SCHEMATICS
          ) {
            continue;
          }
          // TODO: after 4.0: use this test, as we'll have synced versions between lib's and root package.json
          // const expectedVersion = (peerDependencies as Record<
          //   string,
          //   string
          // >)[toAdd];
          const expectedDependency = dependencies[toAdd];
          expect(expectedDependency).toBeTruthy();
          // expect(expectedDependency).toEqual(expectedVersion);
        }
      });
    });
  
    describe('SmartEdit feature', () => {
      describe('general setup', () => {
        beforeEach(async () => {
          appTree = await schematicRunner.runSchematic(
            'ng-add',
            contentfulFeatureOptions,
            appTree
          );
        });
  
        it('should add the feature using the lazy loading syntax', async () => {
          const module = appTree.readContent(contentfulFeatureModulePath);
          expect(module).toMatchSnapshot();
        });
  
        describe('assets', () => {
          it('should update angular.json', async () => {
            const content = appTree.readContent('/angular.json');
            expect(content).toMatchSnapshot();
          });
        });
      });
  
      describe('eager loading', () => {
        beforeEach(async () => {
          appTree = await schematicRunner.runSchematic(
            'ng-add',
            { ...contentfulFeatureOptions, lazy: false },
            appTree
          );
        });
  
        it('should import appropriate modules', async () => {
          const module = appTree.readContent(contentfulFeatureModulePath);
          expect(module).toMatchSnapshot();
        });
      });
  
      describe('with storefrontPreviewRoute config', () => {
        beforeEach(async () => {
          appTree = await schematicRunner.runSchematic(
            'ng-add',
            {
              ...contentfulFeatureOptions,
              storefrontPreviewRoute: 'cx-preview',
            },
            appTree
          );
        });
  
        it('should configure the storefrontPreviewRoute', async () => {
          const module = appTree.readContent(contentfulFeatureModulePath);
          expect(module).toMatchSnapshot();
        });
      });
  
      describe('with allowOrigin config', () => {
        beforeEach(async () => {
          appTree = await schematicRunner.runSchematic(
            'ng-add',
            {
              ...contentfulFeatureOptions,
              allowOrigin: 'localhost:9002',
            },
            appTree
          );
        });
  
        it('should configure the allowOrigin', async () => {
          const module = appTree.readContent(contentfulFeatureModulePath);
          expect(module).toMatchSnapshot();
        });
      });
    });
  });
  