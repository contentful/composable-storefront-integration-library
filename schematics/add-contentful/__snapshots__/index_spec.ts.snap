// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spartacus Contentful schematics: ng-add Contentful feature eager loading should import appropriate modules 1`] = `
"import { NgModule } from '@angular/core';
import { provideConfig } from "@spartacus/core";
import { ContentfulModule } from "@spartacus/contentful";
import { ContentfulConfig, ContentfulRootModule } from "@spartacus/contentful/root";

@NgModule({
  declarations: [],
  imports: [
    ContentfulRootModule,
    ContentfulModule
  ],
  providers: [provideConfig(<ContentfulConfig>{
    contentful: {
      storefrontPreviewRoute: 'STOREFRONT_PREVIEW_ROUTE_PLACEHOLDER',
      allowOrigin: 'ALLOWED_ORIGIN_PLACEHOLDER',
    },
  })]
})
export class ContentfulFeatureModule { }
"
`;

exports[`Spartacus Contentful schematics: ng-add Contentful feature general setup assets should update angular.json 1`] = `
"{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "",
  "projects": {
    "schematics-test": {
      "projectType": "application",
      "schematics": {
        "@schematics/angular:component": {
          "style": "scss",
          "standalone": false
        },
        "@schematics/angular:directive": {
          "standalone": false
        },
        "@schematics/angular:pipe": {
          "standalone": false
        }
      },
      "root": "",
      "sourceRoot": "src",
      "prefix": "app",
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:application",
          "options": {
            "outputPath": "dist/schematics-test",
            "index": "src/index.html",
            "browser": "src/main.ts",
            "polyfills": [
              "zone.js"
            ],
            "tsConfig": "tsconfig.app.json",
            "inlineStyleLanguage": "scss",
            "assets": [
              "src/favicon.ico",
              "src/assets",
              {
                "glob": "**/*",
                "input": "./node_modules/@spartacus/contentful/assets",
                "output": "assets/"
              }
            ],
            "styles": [
              "src/styles.scss"
            ],
            "scripts": [],
            "stylePreprocessorOptions": {
              "includePaths": [
                "node_modules/"
              ]
            }
          },
          "configurations": {
            "production": {
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "500kb",
                  "maximumError": "3.5mb"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "2kb",
                  "maximumError": "4kb"
                }
              ],
              "outputHashing": "all"
            },
            "development": {
              "optimization": false,
              "extractLicenses": false,
              "sourceMap": true
            }
          },
          "defaultConfiguration": "production"
        },
        "serve": {
          "builder": "@angular-devkit/build-angular:dev-server",
          "configurations": {
            "production": {
              "buildTarget": "schematics-test:build:production"
            },
            "development": {
              "buildTarget": "schematics-test:build:development"
            }
          },
          "defaultConfiguration": "development"
        },
        "extract-i18n": {
          "builder": "@angular-devkit/build-angular:extract-i18n",
          "options": {
            "buildTarget": "schematics-test:build"
          }
        },
        "test": {
          "builder": "@angular-devkit/build-angular:karma",
          "options": {
            "polyfills": [
              "zone.js",
              "zone.js/testing"
            ],
            "tsConfig": "tsconfig.spec.json",
            "inlineStyleLanguage": "scss",
            "assets": [
              "src/favicon.ico",
              "src/assets",
              {
                "glob": "**/*",
                "input": "./node_modules/@spartacus/contentful/assets",
                "output": "assets/"
              }
            ],
            "styles": [
              "src/styles.scss"
            ],
            "scripts": [],
            "stylePreprocessorOptions": {
              "includePaths": [
                "node_modules/"
              ]
            }
          }
        }
      }
    }
  }
}"
`;

exports[`Spartacus Contentful schematics: ng-add Contentful feature general setup should add the feature using the lazy loading syntax 1`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, provideConfig } from "@spartacus/core";
import { CONTENTFUL_FEATURE, ContentfulConfig, ContentfulRootModule } from "@spartacus/contentful/root";

@NgModule({
  declarations: [],
  imports: [
    ContentfulRootModule
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      [CONTENTFUL_FEATURE]: {
        module: () =>
          import('@spartacus/contentful').then((m) => m.ContentfulModule),
      },
    }
  }),
  provideConfig(<ContentfulConfig>{
    contentful: {
      storefrontPreviewRoute: 'STOREFRONT_PREVIEW_ROUTE_PLACEHOLDER',
      allowOrigin: 'ALLOWED_ORIGIN_PLACEHOLDER',
    },
  })
  ]
})
export class ContentfulFeatureModule { }
"
`;

exports[`Spartacus Contentful schematics: ng-add Contentful feature with allowOrigin config should configure the allowOrigin 1`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, provideConfig } from "@spartacus/core";
import { CONTENTFUL_FEATURE, ContentfulConfig, ContentfulRootModule } from "@spartacus/contentful/root";

@NgModule({
  declarations: [],
  imports: [
    ContentfulRootModule
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      [CONTENTFUL_FEATURE]: {
        module: () =>
          import('@spartacus/contentful').then((m) => m.ContentfulModule),
      },
    }
  }),
  provideConfig(<ContentfulConfig>{
    contentful: {
      storefrontPreviewRoute: 'STOREFRONT_PREVIEW_ROUTE_PLACEHOLDER',
      allowOrigin: 'localhost:9002',
    },
  })
  ]
})
export class ContentfulFeatureModule { }
"
`;

exports[`Spartacus Contentful schematics: ng-add Contentful feature with storefrontPreviewRoute config should configure the storefrontPreviewRoute 1`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, provideConfig } from "@spartacus/core";
import { CONTENTFUL_FEATURE, ContentfulConfig, ContentfulRootModule } from "@spartacus/contentful/root";

@NgModule({
  declarations: [],
  imports: [
    ContentfulRootModule
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      [CONTENTFUL_FEATURE]: {
        module: () =>
          import('@spartacus/contentful').then((m) => m.ContentfulModule),
      },
    }
  }),
  provideConfig(<ContentfulConfig>{
    contentful: {
      storefrontPreviewRoute: 'cx-preview',
      allowOrigin: 'ALLOWED_ORIGIN_PLACEHOLDER',
    },
  })
  ]
})
export class ContentfulFeatureModule { }
"
`;
